From 208f499d138aa7311af3eb380df021e3f879e3ef Mon Sep 17 00:00:00 2001
From: sdinkinx <simonx.dinkin@intel.com>
Date: Thu, 27 Jul 2017 15:56:24 +0300
Subject: [PATCH] mac80211: Fix possible sband related NULL pointer

this commit is taken from upstream kernel -
commit 21a8e9dd52 - Thu Apr 27, 2017.
we took the related parts for our use.
this is the commit message:
"Existing API 'ieee80211_get_sdata_band' returns default 2 GHz band even
if the channel context configuration is NULL. This crashes for chipsets
which support 5 Ghz alone when it tries to access members of 'sband'.
Channel context configuration can be NULL in multivif case and when
channel switch is in progress (or) when it fails."

the fix is basically to check that sband is not NULL.

Signed-off-by: sdinkinx <simonx.dinkin@intel.com>
---
 net/mac80211/cfg.c  | 13 ++++++++++---
 net/mac80211/mlme.c |  5 +++++
 2 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index de770d1..f6807a6 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -394,7 +394,8 @@ void sta_set_rate_info_tx(struct sta_info *sta,
 		struct ieee80211_supported_band *sband;
 		sband = sta->local->hw.wiphy->bands[
 				ieee80211_get_sdata_band(sta->sdata)];
-		rinfo->legacy = sband->bitrates[rate->idx].bitrate;
+                if (sband)
+		    rinfo->legacy = sband->bitrates[rate->idx].bitrate;
 	}
 	if (rate->flags & IEEE80211_TX_RC_40_MHZ_WIDTH)
 		rinfo->flags |= RATE_INFO_FLAGS_40_MHZ_WIDTH;
@@ -422,7 +423,8 @@ void sta_set_rate_info_rx(struct sta_info *sta, struct rate_info *rinfo)
 
 		sband = sta->local->hw.wiphy->bands[
 				ieee80211_get_sdata_band(sta->sdata)];
-		rinfo->legacy =
+                if (sband)
+		    rinfo->legacy =
 			sband->bitrates[sta->last_rx_rate_idx].bitrate;
 	}
 
@@ -1189,6 +1191,8 @@ static int sta_apply_parameters(struct ieee80211_local *local,
 	u32 mask, set;
 
 	sband = local->hw.wiphy->bands[band];
+        if (!sband)
+            return -EINVAL;
 
 	mask = params->sta_flags_mask;
 	set = params->sta_flags_set;
@@ -1907,6 +1911,7 @@ static int ieee80211_change_bss(struct wiphy *wiphy,
 				struct bss_parameters *params)
 {
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
+        struct ieee80211_supported_band *sband;
 	enum ieee80211_band band;
 	u32 changed = 0;
 
@@ -1914,6 +1919,9 @@ static int ieee80211_change_bss(struct wiphy *wiphy,
 		return -ENOENT;
 
 	band = ieee80211_get_sdata_band(sdata);
+        sband = wiphy->bands[band];
+        if (!sband)
+            return -EINVAL;
 
 	if (params->use_cts_prot >= 0) {
 		sdata->vif.bss_conf.use_cts_prot = params->use_cts_prot;
@@ -1940,7 +1948,6 @@ static int ieee80211_change_bss(struct wiphy *wiphy,
 	if (params->basic_rates) {
 		int i, j;
 		u32 rates = 0;
-		struct ieee80211_supported_band *sband = wiphy->bands[band];
 
 		for (i = 0; i < params->basic_rates_len; i++) {
 			int rate = (params->basic_rates[i] & 0x7f) * 5;
diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 5194438..80b1c4a 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -2719,6 +2719,11 @@ static bool ieee80211_assoc_success(struct ieee80211_sub_if_data *sdata,
 	}
 
 	sband = local->hw.wiphy->bands[ieee80211_get_sdata_band(sdata)];
+        if (!sband) {
+            mutex_unlock(&sdata->local->sta_mtx);
+            ret = false;
+            goto out;
+        }
 
 	/* Set up internal HT/VHT capabilities */
 	if (elems.ht_cap_elem && !(ifmgd->flags & IEEE80211_STA_DISABLE_HT))
-- 
2.10.1

